clusterous/cluster.py

```diff
@@ -643,6 +640,161 @@ def get_cluster_spec(self):
+                security_group = i
+    
+        if security_group is None:
+            # Private Security group
+            security_group = self.vpc_conn.create_security_group(self.cluster_name+"-{0}".format(sg_name), 'Private security group for ' + self.cluster_name, self.vpc.id)
+            security_group.add_tags(self._tag_resource(sg_name))
+            security_group.authorize(ip_protocol='-1', from_port=0, to_port=65535, cidr_ip='0.0.0.0/0')
+        
+        return security_group
+
+    def _create_public_sg(self, sg_name):
+        security_group = None
+        security_group_name_full = "{0}-{1}".format(self.cluster_name, sg_name)
+        security_groups = self.vpc_conn.get_all_security_groups()
+        for i in security_groups:
+            if i.tags.get('Name','') == security_group_name_full:
+                security_group = i
+    
+        if security_group is None:
+            # Public Security group
+            private_security_group = self._create_private_sg("private-sg")
+            security_group = self.vpc_conn.create_security_group(self.cluster_name+"-{0}".format(sg_name), 'Public security group for ' + self.cluster_name, self.vpc.id)
+            security_group.add_tags(self._tag_resource(sg_name))
+            security_group.authorize(ip_protocol='tcp', from_port=22, to_port=22, cidr_ip='0.0.0.0/0')
+            security_group.authorize(ip_protocol='tcp', from_port=22000, to_port=22000, cidr_ip='0.0.0.0/0')
+            security_group.authorize(ip_protocol='-1', from_port=0, to_port=65535, src_group=private_security_group)
+        
+        return security_group
+
+    def _create_subnet(self, subnet_name):
+        subnet = None
+        subnet_name_full = "{0}-{1}-{2}".format(self.cluster_name, subnet_name, self.default_zone)
+        # Get list of subnets
+        subnets = self.vpc_conn.get_all_subnets(filters={'vpcId':self.vpc.id, 'availabilityZone':self.availability_zone})
+        max_subnet_cidr = 1
+        if subnets:
+            for i in subnets:
+                if i.tags.get('Name','') == subnet_name_full:
+                    subnet = i
+                else:
+                    subnet_cidr = i.cidr_block.split('.')
+                    if int(subnet_cidr[2]) > max_subnet_cidr:
+                        max_subnet_cidr = int(subnet_cidr[2])
+    
+        def check_subnet(cidr):
+            subnet = None
+            try:
+                subnet = self.vpc_conn.create_subnet(self.vpc.id, cidr, availability_zone=self.availability_zone)
+            except boto.exception.EC2ResponseError as e:
+                pass
+            return subnet
+
+        vpc_cidr = self.vpc.cidr_block.split('.')
+        while subnet is None:
+            max_subnet_cidr += 1
+            if max_subnet_cidr > 254:
+                break
+            cidr = '{0}.{1}.{2}.0/24'.format(vpc_cidr[0], vpc_cidr[1], max_subnet_cidr)
+            subnet = check_subnet(cidr)
```

Comment:
This appears to me to be a slow and inefficient way of finding a right CIDR. In the worst case, it would make 253 AWS API calls just for creating a subnet. Isn't there a better way of determining a valid value, e.g. by examining what is currently in use?